libcalico-go/lib/backend/k8s/resources/workloadendpoint.go

```diff
@@ -87,54 +108,75 @@ func (c *WorkloadEndpointClient) Delete(ctx context.Context, key model.Key, revi
 //
 // We store the IP addresses in annotations because patching the PodStatus directly races with changes that
 // kubelet makes so kubelet can undo our changes.
-func (c *WorkloadEndpointClient) patchInAnnotations(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
+func (c *WorkloadEndpointClient) patchInAnnotations(ctx context.Context, kvp *model.KVPair, operation string) (*model.KVPair, error) {
+	var annotations map[string]string
+	var revision string
+	patchMode := PatchModeOf(ctx)
+	switch patchMode {
+	case PatchModeCNI:
+		annotations = c.calcCniAnnotations(kvp)
+		// Note: we drop the revision here because the CNI plugin can't handle a retry right now (and the kubelet
```

Comment:
@fasaxc are you sure this is the right thing to do?

The CNI plugin doesn't have retries, but the CNI interface itself _does_ - the container runtime will retry the CNI call if it hits a conflict. 

Is this race condition prevalent enough that we need this?

Reply:
And if so, we should probably be adding retry logic to the CNI plugin rather than hacking around it and ignoring the race.

Reply:
This was pre-existing code before the PR.
If you guys determine that the CNI Plugin should have retry logic, I'd be more comfortable if someone more familiar with the CNI Plugin code made that change in a separate PR.

Reply:
Sorry I'm not spotting it - where in the pre-existing code did we set the revision to an empty string?

From what I can tell, the previous code ensured that we always respected the provided revision string.

My understanding is that we should _never_ secretly drop the revision that was provided by the calling code. If the caller provided a revision, we NEED to respect that. If we think the CNI plugin shouldn't be passing a revision, we should probably change the arguments that the CNI plugin itself passes to `Update` so that it doesn't pass a revision.

Reply:
It's hard to see since I've changed the structure of the code a bit. This [link](https://github.com/projectcalico/calico/blob/master/libcalico-go/lib/backend/k8s/resources/workloadendpoint.go#L100-L102) to master shows that Create and Update always passed the revision as "" which I found strange.
If you think I'm correct on that, I can implement your suggestion above to ensure the calling code never passes the revision down since we're not respecting it anyway?

Reply:
Ah yeah, thanks for the pointer.

No, I don't want to burden you with that refactor now, and I'd want to think more about whether or not it has knock-on consequences.

The right fix is still to add retry logic to the CNI plugin and properly set the revision... but that's potentially complicated bit of work.

Reply:
Raised this to track it: https://github.com/projectcalico/calico/issues/5707